tracing-tracy/src/lib.rs

```diff
@@ -228,55 +257,86 @@ where
                     self.stack_depth,
                 ),
                 id.into_u64(),
-            ));
+            )
+        };
+
+        TRACY_SPAN_STACK.with(|s| {
+            s.push(stack_frame);
         });
     }
 
     fn on_exit(&self, id: &Id, _: Context<S>) {
-        TRACY_SPAN_STACK.with(|s| {
-            if let Some((span, span_id)) = s.borrow_mut().pop_back() {
-                if id.into_u64() != span_id {
-                    self.client.color_message(
-                        "Tracing spans exited out of order! \
-                        Trace may not be accurate for this span stack.",
-                        0xFF000000,
-                        self.stack_depth,
-                    );
-                }
-                drop(span);
-            } else {
+        let stack_frame = TRACY_SPAN_STACK.with(VecCell::pop);
+
+        if let Some((span, span_id)) = stack_frame {
+            if id.into_u64() != span_id {
                 self.client.color_message(
-                    "Exiting a tracing span, but got nothing on the tracy span stack!",
+                    "Tracing spans exited out of order! \
+                        Trace may not be accurate for this span stack.",
                     0xFF000000,
                     self.stack_depth,
                 );
             }
-        });
+            drop(span);
+        } else {
+            self.client.color_message(
+                "Exiting a tracing span, but got nothing on the tracy span stack!",
+                0xFF000000,
+                self.stack_depth,
+            );
+        }
+    }
+
+    fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+        let Some(span) = ctx.span(&id) else { return };
+
+        if let Some(fields) = span.extensions_mut().get_mut::<FormattedFields<F>>() {
+            let buf = mem::take(&mut fields.fields);
+            CACHE.with(|cache| drop(StrCacheGuard::new(cache, buf)));
+        };
     }
 }
 
-struct TracyEventFieldVisitor {
-    dest: String,
+struct TracyEventFieldVisitor<'a> {
+    dest: &'a mut String,
     frame_mark: bool,
     first: bool,
 }
 
-impl Visit for TracyEventFieldVisitor {
+impl Visit for TracyEventFieldVisitor<'_> {
     fn record_bool(&mut self, field: &Field, value: bool) {
         match (value, field.name()) {
-            (true, "tracy.frame_mark") => self.frame_mark = true,
-            _ => self.record_debug(field, &value),
+            (_, "tracy.frame_mark") => self.frame_mark = value,
+            (true, _) => self.record_str(field, "true"),
+            (false, _) => self.record_str(field, "false"),
         }
     }
 
+    fn record_str(&mut self, field: &Field, value: &str) {
+        {
```

Comment:
Why the extra block? I would also extract `field.name()` into a variable.

Reply:
Trying to scope the `alloc_always_size` variable, but eh kinda unnecessary. Done.